<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.433">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>HAB796B9: Software development for data sciences - 2023-2024 Project</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="../../custom.scss">
</head>

<body class="nav-sidebar floating nav-fixed fullcontent">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">HAB796B9</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-general-tools" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">General tools</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-general-tools">    
        <li>
    <a class="dropdown-item" href="../../Courses/Bash/tp.html" rel="" target="">
 <span class="dropdown-text">Command-line tools</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../Courses/Git/tp.html" rel="" target="">
 <span class="dropdown-text">Version control with Git</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../Courses/IDE/tp.html" rel="" target="">
 <span class="dropdown-text">Integrated Development Environment</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../Courses/CI/tp.html" rel="" target="">
 <span class="dropdown-text">Continuous Integration</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../Courses/Docs/tp.html" rel="" target="">
 <span class="dropdown-text">Markup languages / Documentation</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-python" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Python</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-python">    
        <li class="dropdown-header">Dev. Tools</li>
        <li>
    <a class="dropdown-item" href="../../Courses/Venv/tp.html" rel="" target="">
 <span class="dropdown-text">Virtual Python Environment</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../Courses/ClassesExceptions/tp.html" rel="" target="">
 <span class="dropdown-text">Classes &amp; Exceptions</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../Courses/Python-modules/tp.html" rel="" target="">
 <span class="dropdown-text">Creating a <code>python</code> module</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../Courses/Test/tp.html" rel="" target="">
 <span class="dropdown-text">Unit Tests</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../Courses/ProfilingDebugging/tp.html" rel="" target="">
 <span class="dropdown-text">Debugging &amp; Profiling</span></a>
  </li>  
        <li><hr class="dropdown-divider"></li>
        <li class="dropdown-header">Scientific Python</li>
        <li>
    <a class="dropdown-item" href="../../Courses/Numpy/tp.html" rel="" target="">
 <span class="dropdown-text">Numpy cheat sheet</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../Courses/ScipyNumpy/tp.html" rel="" target="">
 <span class="dropdown-text">SciPy</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../Courses/TimeMemory/tp.html" rel="" target="">
 <span class="dropdown-text">Sparse matrices, graphs and maps</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../Courses/Visualization/tp.html" rel="" target="">
 <span class="dropdown-text">Data Visualization</span></a>
  </li>  
        <li><hr class="dropdown-divider"></li>
        <li class="dropdown-header">Pandas</li>
        <li>
    <a class="dropdown-item" href="../../Courses/Pandas/pandas_titanic.html" rel="" target="">
 <span class="dropdown-text">Titanic dataset</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../Courses/Pandas/pandas_airparif.html" rel="" target="">
 <span class="dropdown-text">Airparif dataset</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../Courses/Pandas/pandas_bikes.html" rel="" target="">
 <span class="dropdown-text">Bike accident dataset</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-grading" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Grading</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-grading">    
        <li>
    <a class="dropdown-item" href="../../Projects/2023-2024/README.html" rel="" target="">
 <span class="dropdown-text">2023-2024 Project</span></a>
  </li>  
    </ul>
  </li>
</ul>
            <div class="quarto-navbar-tools ms-auto">
    <a href="https://moodle.umontpellier.fr/course/view.php?id=28921" rel="" title="Moodle" class="quarto-navigation-tool px-1" aria-label="Moodle"><i class="bi bi-bicycle"></i></a>
    <a href="https://github.com/bcharlier/HAB796B9" rel="" title="GitHub" class="quarto-navigation-tool px-1" aria-label="GitHub"><i class="bi bi-github"></i></a>
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../Projects/2023-2024/README.html">2023-2024 Project</a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../Projects/2023-2024/README.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">2023-2024 Project</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">2023-2024 Project</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<section id="guidelines" class="level2">
<h2 class="anchored" data-anchor-id="guidelines">Guidelines</h2>
<section id="schedule" class="level3">
<h3 class="anchored" data-anchor-id="schedule">Schedule</h3>
<p>The project should be done in groups of 4 peoples. The project’s deadline is <em>December 12, 2023</em>, from 13:15 to 14:45 in room TD5.127. Each group will be required to make an oral presentation.</p>
<p>You will have the opportunity to ask questions about the project on <em>November 21, 2023</em>, from 13:15 to 14:45 in room TD36.209. Groups should be formed on this date, and various tasks should be assigned to team members.</p>
</section>
<section id="structure" class="level3">
<h3 class="anchored" data-anchor-id="structure">Structure</h3>
<p>The primary goals of this project are as follows:</p>
<ul>
<li><p>a GitHub repository containing all the code and description of your project.</p></li>
<li><p>Organize your project structure as illustrated below, with <code>pygen</code> serving as the project’s name:</p></li>
</ul>
<div class="sourceCode" id="cb1"><pre class="sourceCode default code-with-copy"><code class="sourceCode default"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>project_root_dir/</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>   ├── pygen/</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>   │     ├── __init__.py</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>   │     ├── words/</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>   │     │     └── __init__.py</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>   │     └── convert/</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>   │           └── __init__.py</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>   ├── Readme.md</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a>   └── .gitignore</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p>The <code>Readme.md</code> file should include a comprehensive project description and provide a minimal working example that demonstrates the functionality of the code.</p></li>
<li><p>Create slides for a brief presentation lasting 15 minutes, with an additional 5 minutes allocated for questions. The slide file should be included in the project repository.</p></li>
</ul>
<p>These objectives will help ensure the successful completion and presentation of your project.</p>
</section>
<section id="grading" class="level3">
<h3 class="anchored" data-anchor-id="grading">Grading</h3>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Important
</div>
</div>
<div class="callout-body-container callout-body">
<p>The project repository <strong>must show a balanced contribution between group members</strong> and intra-group grade variation could be made to reflect issues on the intra-group workload balance.</p>
</div>
</div>
<table class="table-striped table">
<caption></caption>
<thead>
<tr class="header">
<th>General</th>
<th>Details</th>
<th>Points</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><strong>Git</strong></td>
<td>Git repo (branch, commit, …)</td>
<td>3</td>
</tr>
<tr class="even">
<td><strong>Code</strong></td>
<td>Structure of the code</td>
<td>1</td>
</tr>
<tr class="odd">
<td></td>
<td>Description / Documentation</td>
<td>2</td>
</tr>
<tr class="even">
<td><strong>Work description</strong></td>
<td>Clarity / Details</td>
<td>2</td>
</tr>
<tr class="odd">
<td></td>
<td>Answer of Projects goals</td>
<td>8</td>
</tr>
<tr class="even">
<td><strong>Oral</strong></td>
<td>Slides</td>
<td>2</td>
</tr>
<tr class="odd">
<td></td>
<td>Oral presentation</td>
<td>2</td>
</tr>
<tr class="even">
<td><strong>Total</strong></td>
<td></td>
<td>20</td>
</tr>
</tbody>
</table>
<p>The Git component will assess your group’s capacity to collaboratively merge individual contributions. Individuals who have not made any contributions in Git will incur a penalty.</p>
</section>
</section>
<section id="english-words-in-the-human-proteome" class="level1">
<h1>English Words in the Human Proteome</h1>
<p>The objective of this first project is to discover if English words can be found in the sequences of the human proteome, i.e., in the sequences of all human proteins.</p>
<section id="amino-acid-composition" class="level3">
<h3 class="anchored" data-anchor-id="amino-acid-composition">Amino Acid Composition</h3>
<p>In the first step, create 5 English words using the 20 amino acids.</p>
</section>
<section id="words" class="level3">
<h3 class="anchored" data-anchor-id="words">Words</h3>
<ol type="1">
<li><p>Download the file <a href="https://python.sdv.univ-paris-diderot.fr/data-files/english-common-words.txt"><code>english-common-words.txt</code></a>. This file contains the 3000 most common English words, with one word per line.</p></li>
<li><p>Create a script <code>words_in_proteome.py</code> and write the function <code>read_words()</code> to read the words from the file provided as an argument to the script and return a list containing the words converted to uppercase and composed of 3 characters or more.</p></li>
<li><p>In the main program, display the number of selected words.</p></li>
</ol>
</section>
<section id="proteins" class="level3">
<h3 class="anchored" data-anchor-id="proteins">Proteins</h3>
<ol type="1">
<li><p>Download the file <a href="https://python.sdv.univ-paris-diderot.fr/data-files/human-proteome.fasta"><code>human-proteome.fasta</code></a>. Note that this file is quite large and comes from the UniProt database from this <a href="https://www.uniprot.org/help/human_proteome">page</a>.</p></li>
<li><p>Here are the first lines of this file (the <code>[...]</code> indicates a cut we made):</p></li>
</ol>
<div class="sourceCode" id="cb2"><pre class="sourceCode default code-with-copy"><code class="sourceCode default"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>    sp|O95139|NDUB6_HUMAN NADH dehydrogenase [ubiquinone] 1 beta [...]</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>    MTGYTPDEKLRLQQLRELRRRWLKDQELSPREPVLPPQKMGPMEKFWNKFLENKSPWRKM</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>    VHGVYKKSIFVFTHVLVPVWIIHYYMKYHVSEKPYGIVEKKSRIFPGDTILETGEVIPPM</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>    KEFPDQHH</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>    sp|O75438|NDUB1_HUMAN NADH dehydrogenase [ubiquinone] 1 beta [...]</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>    MVNLLQIVRDHWVHVLVPMGFVIGCYLDRKSDERLTAFRNKSMLFKRELQPSEEVTWK</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>    sp|Q8N4C6|NIN_HUMAN Ninein OS=Homo sapiens OX=9606 GN=NIN PE=1 SV=4</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>    MDEVEQDQHEARLKELFDSFDTTGTGSLGQEELTDLCHMLSLEEVAPVLQQTLLQDNLLG</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>    RVHFDQFKEALILILSRTLSNEEHFQEPDCSLEAQPKYVRGGKRYGRRSLPEFQESVEEF</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>    PEVTVIEPLDEEARPSHIPAGDCSEHWKTQRSEEYEAEGQLRFWNPDDLNASQSGSSPPQ</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ol start="3" type="1">
<li><p>In the <code>words_in_proteome.py</code> script, write the function <code>read_sequences()</code> that will read the proteome from the file provided as the second argument of the script. This function will return a dictionary with protein identifiers as keys (e.g., O95139, O75438, Q8N4C6) and their associated sequences as values.</p></li>
<li><p>In the main program, display the number of sequences read. For testing purposes, also display the sequence associated with the protein O95139.</p></li>
</ol>
</section>
<section id="searching-for-words" class="level3">
<h3 class="anchored" data-anchor-id="searching-for-words">Searching for Words</h3>
<ol type="1">
<li>Write the function <code>search_words_in_proteome()</code> that takes as arguments the list of words and the dictionary containing protein sequences. This function will count the number of sequences in which a word is present. The function will return a dictionary with words as keys and the number of sequences containing these words as values. The function will also display the following message for words found in the proteome:</li>
</ol>
<div class="sourceCode" id="cb3"><pre class="sourceCode default code-with-copy"><code class="sourceCode default"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>ACCESS found in 1 sequences</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>ACID found in 38 sequences</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>ACT found in 805 sequences</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>This step may take a few minutes. Please be patient.</p>
</section>
<section id="the-most-frequent-word" class="level3">
<h3 class="anchored" data-anchor-id="the-most-frequent-word">The Most Frequent Word</h3>
<ol type="1">
<li>Finally, write the function <code>find_most_frequent_word()</code> that takes the dictionary returned by the previous function <code>search_words_in_proteome()</code> as an argument. This function will display the word found in the most sequences and the number of sequences in which it was found in the following format:</li>
</ol>
<div class="sourceCode" id="cb4"><pre class="sourceCode default code-with-copy"><code class="sourceCode default"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>=&gt; xxx found in yyy sequences</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ol start="2" type="1">
<li><p>What is this word?</p></li>
<li><p>What percentage of the proteome sequences contain this word?</p></li>
</ol>
</section>
<section id="being-more-comprehensive" class="level3">
<h3 class="anchored" data-anchor-id="being-more-comprehensive">Being More Comprehensive</h3>
<p>Up to this point, we have determined, for each word, the number of sequences in which it appears. We could go further and also calculate how many times each word appears in the sequences.</p>
<ol type="1">
<li><p>To do this, modify the <code>search_words_in_proteome()</code> function to count the number of occurrences of a word in the sequences. The <code>.count()</code> method will be useful.</p></li>
<li><p>Determine which word is the most frequent in the human proteome.</p></li>
</ol>
</section>
</section>
<section id="genbank-to-fasta-converter" class="level1">
<h1>GenBank to FASTA Converter</h1>
<p>This project involves creating a file converter from GenBank format to FASTA format. The web page <a href="https://python.sdv.univ-paris-diderot.fr/annexe_formats_fichiers/">Some Data Formats Encountered in Biology</a> provides details about these two file formats.</p>
<p>The dataset we will work with is the GenBank file of the yeast Saccharomyces cerevisiae’s chromosome I. The original file may be found at this <a href="https://www.ncbi.nlm.nih.gov/nuccore/NC_001133">url</a>.</p>
<section id="reading-the-file" class="level3">
<h3 class="anchored" data-anchor-id="reading-the-file">Reading the File</h3>
<ol type="1">
<li><p>Create a script <code>genbank2fasta.py</code> and write the function <code>read_file()</code> that takes the file’s name as an argument and returns the file’s content as a list of lines, where each line is a string.</p></li>
<li><p>Test this function with the GenBank file <a href="https://python.sdv.univ-paris-diderot.fr/data-files/NC_001133.gbk"><code>NC_001133.gbk</code></a> and display the number of lines read.</p></li>
</ol>
</section>
<section id="extracting-the-organism-name" class="level3">
<h3 class="anchored" data-anchor-id="extracting-the-organism-name">Extracting the Organism Name</h3>
<p>In the same script, add the function <code>extract_organism()</code> that takes the content of the previously obtained file (as a list of lines) with the <code>read_file()</code> function and returns the name of the organism. To retrieve the correct line, you can check if the line’s first characters contain the keyword <code>ORGANISM</code>.</p>
<ol start="3" type="1">
<li>Test this function with the GenBank file <a href="https://python.sdv.univ-paris-diderot.fr/data-files/NC_001133.gbk"><code>NC_001133.gbk</code></a> and display the organism’s name.</li>
</ol>
</section>
<section id="finding-genes" class="level3">
<h3 class="anchored" data-anchor-id="finding-genes">Finding Genes</h3>
<p>In the GenBank file, sense genes are indicated like this:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode default code-with-copy"><code class="sourceCode default"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>   gene            58..272</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>or</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode default code-with-copy"><code class="sourceCode default"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>   gene            &lt;2480..&gt;2707</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>and antisense genes (or complementary) like this:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode default code-with-copy"><code class="sourceCode default"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>   gene            complement(55979..56935)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>or</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode default code-with-copy"><code class="sourceCode default"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>   gene            complement(&lt;13363..&gt;13743)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>The numeric values separated by <code>..</code> indicate the gene’s position in the genome (first base number, last base number).</p>
<div class="callout callout-style-default callout-important no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Important
</div>
</div>
<div class="callout-body-container callout-body">
<p>The <code>&lt;</code> symbol indicates a partial gene at the 5’ end, meaning that the corresponding START codon is incomplete. Similarly, the <code>&gt;</code> symbol denotes a partial gene at the 3’ end, indicating that the corresponding STOP codon is incomplete. For more details, refer to the <a href="https://www.ncbi.nlm.nih.gov/Sitemap/samplerecord.html#BaseSpanB">NCBI documentation on gene boundaries</a>. Here, we propose to ignore these <code>&gt;</code> and <code>&lt;</code> symbols.</p>
</div>
</div>
<p>Locate these different genes in the <a href="https://python.sdv.univ-paris-diderot.fr/data-files/NC_001133.gbk"><code>NC_001133.gbk</code></a> file. To retrieve these gene lines, check if the line begins with</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode default code-with-copy"><code class="sourceCode default"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>     gene            </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>(i.e., 5 spaces, followed by the word “gene,” followed by 12 spaces). To determine if it’s a sense or antisense gene, check for the presence of the word <code>complement</code> in the line read.</p>
<ol start="4" type="1">
<li>Then, if you wish to retrieve the start and end positions of the gene, we recommend using the <code>replace()</code> function and keeping only the numbers and <code>..</code>. For example,</li>
</ol>
<div class="sourceCode" id="cb10"><pre class="sourceCode default code-with-copy"><code class="sourceCode default"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>    gene            &lt;2480..&gt;2707</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>will be changed to</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode default code-with-copy"><code class="sourceCode default"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>2480..2707</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Finally, using the <code>.split()</code> method, you can easily retrieve the two integer values for the start and end of the gene.</p>
<ol start="5" type="1">
<li>In the same <code>genbank2fasta.py</code> script, add the function <code>find_genes()</code> that takes the file’s content (as a list of lines) as an argument and returns a list of genes.</li>
</ol>
<p>Each gene will itself be a list containing the first base number, the last base number, and a string “sense” for a sense gene and “antisense” for an antisense gene.</p>
<ol start="6" type="1">
<li>Test this function with the GenBank file <a href="https://python.sdv.univ-paris-diderot.fr/data-files/NC_001133.gbk"><code>NC_001133.gbk</code></a> and display the number of genes found, as well as the number of sense and antisense genes.</li>
</ol>
</section>
<section id="extracting-the-nucleotide-sequence-of-the-genome" class="level3">
<h3 class="anchored" data-anchor-id="extracting-the-nucleotide-sequence-of-the-genome">Extracting the Nucleotide Sequence of the Genome</h3>
<p>The genome size is indicated on the first line of a GenBank file. Find the genome size stored in the <a href="https://python.sdv.univ-paris-diderot.fr/data-files/NC_001133.gbk"><code>NC_001133.gbk</code></a> file.</p>
<p>In a GenBank file, the genome’s sequence is located between the lines</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode default code-with-copy"><code class="sourceCode default"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>ORIGIN</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>and</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode default code-with-copy"><code class="sourceCode default"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>//</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ol type="1">
<li><p>Find the first and last lines of the genome sequence in the <code>NC_001133.gbk</code> file.</p></li>
<li><p>To retrieve the lines containing the sequence, we suggest using an algorithm with a flag <code>is_dnaseq</code> (which will be either <code>True</code> or <code>False</code>). Here is the proposed pseudo-code algorithm:</p></li>
</ol>
<div class="sourceCode" id="cb14"><pre class="sourceCode default code-with-copy"><code class="sourceCode default"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>is_dnaseq &lt;- False</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>Read each line of the gbk file</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>    If the line contains "//"</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>        is_dnaseq &lt;- False</span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>    If is_dnaseq is True</span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>        Accumulate the sequence</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>    If the line contains "ORIGIN"</span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>        is_dnaseq &lt;- True</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>At the beginning, this flag will have the value <code>False</code>. Then, when it becomes <code>True</code>, you can read the lines containing the sequence, and when it becomes <code>False</code> again, you will stop.</p>
<ol start="3" type="1">
<li>Once the sequence is retrieved, simply remove the numbers, carriage returns, and other spaces (Tip: calculate the length of the sequence and compare it to the one mentioned in the gbk file).</li>
</ol>
<p>Still in the same <code>genbank2fasta.py</code> script, add the function <code>extract_sequence()</code> that takes the file’s content (as a list of lines) as an argument and returns the nucleotide sequence of the genome (as a string). The sequence should not contain spaces, numbers, or carriage returns.</p>
<ol start="7" type="1">
<li>Test this function with the GenBank file <a href="https://python.sdv.univ-paris-diderot.fr/data-files/NC_001133.gbk"><code>NC_001133.gbk</code></a> and display the number of bases in the extracted sequence. Verify that you have not made any errors by comparing the size of the extracted sequence with the one found in the GenBank file.</li>
</ol>
</section>
<section id="creating-the-reverse-complementary-sequence" class="level3">
<h3 class="anchored" data-anchor-id="creating-the-reverse-complementary-sequence">Creating the Reverse Complementary Sequence</h3>
<p>Still in the same script, add the function <code>construct_comp_inverse()</code> that takes a DNA sequence as a string and returns the reverse complementary sequence (also as a string).</p>
<p>Remember that constructing the reverse complementary sequence of a DNA sequence involves:</p>
<ul>
<li>Taking the complementary sequence, which means replacing base <code>a</code> with base <code>t</code>, <code>t</code> with <code>a</code>, <code>c</code> with <code>g</code>, and <code>g</code> with <code>c</code>.</li>
<li>Taking the reverse, which means that the first base of the complementary sequence becomes the last base and vice versa, and so on.</li>
</ul>
<p>To make your work easier, only work with lowercase sequences.</p>
<ol start="8" type="1">
<li>Test this function with the sequences <code>atcg</code>, <code>AATTCCGG</code>, and <code>gattaca</code>.</li>
</ol>
</section>
<section id="writing-a-fasta-file" class="level3">
<h3 class="anchored" data-anchor-id="writing-a-fasta-file">Writing a FASTA File</h3>
<p>Still in the same script, add the function <code>write_fasta()</code> that takes a file name (as a string), a comment (as a string), and a sequence (as a string) as arguments, and writes a FASTA file. The sequence should be written in lines no longer than 80 characters.</p>
<p>As a reminder, a FASTA file follows the following format:</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode default code-with-copy"><code class="sourceCode default"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>&gt;comment</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>sequence on a line with a maximum of 80 characters</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>sequence continuation .........................</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>sequence continuation .........................</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>...</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ol start="9" type="1">
<li><p>Test this function with the following:</p>
<ul>
<li>File name: <code>test.fasta</code></li>
<li>Comment: <code>my comment</code></li>
<li>Sequence: <code>atcgatcgatcgatcgatcgatcgatcgatcgatcgatcgatcgatcgatcgatcgatcgatcgatcgatcgatcgatcgatcgatcgatcgatcg</code></li>
</ul></li>
</ol>
</section>
<section id="extracting-genes" class="level3">
<h3 class="anchored" data-anchor-id="extracting-genes">Extracting Genes</h3>
<p>Still in the same script, add the function <code>extract_genes()</code> that takes the list of genes, the complete nucleotide sequence (as a string), and the organism’s name (as a string) as arguments. For each gene:</p>
<ul>
<li>Extract the gene’s sequence in the complete sequence.</li>
<li>Take the reverse complementary sequence (using the <code>construct_comp_inverse()</code> function if the gene is antisense).</li>
<li>Save the gene in a FASTA format file (using the <code>write_fasta()</code> function).</li>
<li>Display on the screen the gene number and the name of the created FASTA file.</li>
</ul>
<p>The first line of the FASTA files will be in the following format:</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode default code-with-copy"><code class="sourceCode default"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>&gt;organism-name|gene-number|start|end|sense or antisense</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>The gene number will be consecutive from the first gene to the last. There will be no numbering difference between sense and antisense genes.</p>
<ol start="10" type="1">
<li>Test this function with the GenBank file <code>NC_001133.gbk</code>.</li>
</ol>
</section>
<section id="building-the-final-script" class="level3">
<h3 class="anchored" data-anchor-id="building-the-final-script">Building the Final Script</h3>
<p>To finish, modify the <code>genbank2fasta.py</code> script so that the GenBank file to be analyzed (in this example, <code>NC_001133.gbk</code>) is entered as an argument of the script.</p>
<p>You will display an error message if:</p>
<ul>
<li>The <code>genbank2fasta.py</code> script is used without an argument.</li>
<li>The file provided as an argument does not exist.</li>
</ul>
<p>To help you, you may read the documentation of the <code>sys</code> and <code>os</code> modules.</p>
<p>Test your finalized script.</p>


</section>
</section>

<a onclick="window.scrollTo(0, 0); return false;" role="button" id="quarto-back-to-top"><i class="bi bi-arrow-up"></i> Back to top</a></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



<footer class="footer"><div class="nav-footer"><div class="nav-footer-center"><div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.com/bcharlier/HAB796B9/blob/main/Projects/2023-2024/README.qmd" class="toc-action">View source</a></p><p><a href="https://github.com/bcharlier/HAB796B9/issues/new" class="toc-action">Report an issue</a></p></div></div></div></div></footer></body></html>